New doc: Order Form Sync Feature (#36)
vtex/faststore-docs: docs/how-to-guides/platform-integration/vtex/implementing-order-form-sync.md

```tsx
        import styles from 'src/components/ui/Button/button.module.scss'
        import Icon from 'src/components/ui/Icon'
        import { useSession } from 'src/sdk/session'
        import { useCart } from '../../../../sdk/cart/index'

        const ButtonSignIn = () => {
        const { id } = useCart()
        const { person } = useSession()

        return (
            <LinkButton
            data-fs-button-signin-link
            href={
                person?.id ? `/account?orderFormId=${id}` : `/login?orderFormId=${id}`
            }
            className={`${styles.fsButton} text__title-mini`}
            variant="tertiary"
            icon={<Icon name="User" width={18} height={18} weight="bold" />}
```
#### `src/pages/account.tsx` 
Responsible for rendering the My Account page. Add the following to allow subsequent pages to reference the user’s session cookie on this page:

```tsx
        function Page() {
            useEffect(() => {

>             +  window.location.href = `${storeConfig.accountUrl}${window.location.search}`

Reviewer: ```suggestion
              window.location.href = `${storeConfig.accountUrl}${window.location.search}`
```

            }, [])

            return (
```

#### `src/pages/login.tsx` 
Responsible for rendering the Login page. Add the following to allow subsequent pages to reference the user’s session cookie on this page:

```tsx
        function Page() {
        useEffect(() => {

>         +    window.location.href = +`${storeConfig.loginUrl}${window.location.search}`

Reviewer: ```suggestion
            window.location.href = +`${storeConfig.loginUrl}${window.location.search}`
```

---
title: "Implementing Order Form Sync"
sidebar_label: "Implementing Order Form Sync"
---

The Order Form Sync feature maintains the user's session and order information across different pages, such as My Account and Checkout, in the FastStore environment. 

Pages like My Account, Login, and Checkout are in the Legacy Portal (vtexcommercestable) by default. Since FastStore is the main domain and pages in the legacy portal are subdomains, the user's session cookies for these domains must be synchronized to maintain a seamless shopping experience without losing progress or data.

> This synchronization is achieved by duplicating the session cookie for the main domain (faststore) whenever the user navigates to the portal (secure.{dominio}), and a service is called to update the Order Form information associated with that session. This ensures that the user's order data is synchronized across different pages.

Reviewer: ```suggestion
This synchronization is achieved by duplicating the session cookie for the main domain (faststore) whenever the user navigates to the portal (secure.<domain>), and a service is called to update the Order Form information associated with that session. This ensures that the user's order data is synchronized across different pages.
```

:::

---

## Step by step

### Step 1: Add a script to the Checkout

Add the function below to the `checkout6-custom.js` file in the Admin Checkout to replicate the session cookies at the cart and call the endpoint of the service responsible for this replication:

1. In the VTEX Admin, access **Store Settings > STOREFRONT > Checkout**.
2. In **Checkout**, click on `Edit` in the **Default** Website.
3. Navigate to the tab **Code** and in **Files**, click on `checkout6-custom.js` and add the following function:

```JavaScript
function callUrlWithRetry(url, maxRetries, retryInterval) {
  return fetch(url,{
    method: 'POST',
  })
    .then(response => {
      if (response.ok) {
        return 
      } else {
        throw new Error(`Without vtex_session cookie`);
      }
    })
    .catch(error => {
      if (maxRetries > 0) {
        return new Promise(resolve => {
          setTimeout(() => {
            resolve(callUrlWithRetry(url, maxRetries - 1, retryInterval));
          }, retryInterval);
        });
      } else {
        throw error;
      }
    });
}

callUrlWithRetry(`/api/io/faststore/cookie`, 3, 500)
```
4. Click on `Save`.

:::caution 

> It is important to perform a retry as, the session cookie may not be immediately available in the request even after the script has run.

Reviewer: ```suggestion
It is important to perform a retry as the session cookie may not be immediately available in the request even after the script has run.
```

:::

### Step 2: Update FastStore Theme Code

> To send the `orderFormId` parameter, you must perform changes in the FastStore theme code. In your FastStore theme code, navigate to following files:

Reviewer: ```suggestion
To send the `orderFormId` parameter, you must perform changes in the FastStore starter code. Navigate to following files:
```

#### `src/components/ui/Button/ButtonSignIn/ButtonSignIn.tsx`

Responsible for rendering the `Sign In` button on the Checkout page and provides the current cart information. Add the `orderFormId` parameter to allow subsequent pages to reference the order form associated with the cart for tracking and processing purposes:
 
```JavaScritpt
    import styles from 'src/components/ui/Button/button.module.scss'
    import Icon from 'src/components/ui/Icon'
    import { useSession } from 'src/sdk/session'
  + import { useCart } from '../../../../sdk/cart/index'

    const ButtonSignIn = () => {
    const { id } = useCart()
    const { person } = useSession()

    return (
        <LinkButton
        data-fs-button-signin-link
        +  href={
        +    person?.id ? `/account?orderFormId=${id}` : +`/login?orderFormId=${id}`
        +  }
        className={`${styles.fsButton} text__title-mini`}
        variant="tertiary"
        icon={<Icon name="User" width={18} height={18} weight="bold" />}

```
#### `src/pages/account.tsx` 

> Responsible for rendering the My Account page. Add the following to allow subsequent pages to reference the the user’s session cookie of this page:

Reviewer: ```suggestion
Responsible for rendering the My Account page. Add the following to allow subsequent pages to reference the user’s session cookie on this page:
```

```JavaScript

function Page() {
  useEffect(() => {
  +  window.location.href = `${storeConfig.accountUrl}${window.location.search}`
  }, [])

  return (

```

#### `src/pages/login.tsx` 

> Responsible for rendering the Login page. Add the following to allow subsequent pages to reference the the user’s session cookie of this page:

Reviewer: ```suggestion
Responsible for rendering the Login page. Add the following to allow subsequent pages to reference the user’s session cookie on this page:
```

            return;
          } else {
            throw new Error(`Without vtex_session cookie`);
          }
        })
        .catch((error) => {
          if (maxRetries > 0) {
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve(callUrlWithRetry(url, maxRetries - 1, retryInterval));
              }, retryInterval);
            });
          } else {
            throw error;
          }
        });
    }
  
    callUrlWithRetry(`/api/io/faststore/cookie`, 3, 5000);
    if (orderId) {
      callUrlWithRetry(
        `/api/checkout/pub/orderForm/${orderId}?refreshOutdatedData=true`,
        3,
        5000
      );
    }
  </script>

```
## Creating Session Cookie in FastStore Store for Localhost or Internal URLs
The service app generates the session cookie on the main domain. Still, if the store is using localhost or an internal URL at the checkout, it is necessary to create the cookie manually in the faststore store. To do so, navigate to the checkout URL and copy the cookie value. For example:

![session-cookie-example](https://user-images.githubusercontent.com/67270558/232832503-e1426219-caa2-487d-a220-1c26a8226bbf.png)

> Go back to the Fast Store store and use the following command to generate the cookie that will be used to sync the `orderForm`:

Reviewer: Where we might run this command? Devtools console?  Would be good explicit it :)

I noticed different patterns: `Faststore`, `Fast Store`, `FastStore`, `faststore`. WDYT about keeping the same pattern?

Author: Good point! Just added about the console and also follow the pattern as `FastStore`.

```JavaScript
document.cookie = "vtex_session={{value of session cookie}}";

```

> After that, you should see the cookie created and the order form synced.

Reviewer: ⛏️ Nitpicking: the same thing related to the pattern: `order form`, `OrderForm`, `orderForm`.
